runtime/src/consensus/tendermint/verifier.rs

```diff
@@ -0,0 +1,686 @@
+        if runtime_header.namespace != self.trust_root.runtime_id {
+            return Err(Error::VerificationFailed(anyhow!(
+                "header namespace does not match trusted runtime id"
+            )));
+        }
+
+        if runtime_header.round < cache.last_verified_round {
+            // Ignore requests for earlier rounds.
+            return Err(Error::VerificationFailed(anyhow!(
+                "round seems to have moved backwards"
+            )));
+        }
+
+        // Verify the consensus layer block first to obtain an authoritative state root.
+        let consensus_block = self.verify_consensus_only(cache, instance, consensus_block)?;
+        let state_root = consensus_block.get_state_root();
+        let state = ConsensusState::from_protocol(self.protocol.clone(), state_root);
+
+        // Check if we have already verified this runtime header to avoid re-verification.
+        if let Some(state_root) = cache.verified_state_roots.get(&runtime_header.round) {
+            if state_root == &runtime_header.state_root {
+                // Header matches, no need to perform re-verification.
+                return Ok(state);
+            }
+
+            // Header is for the same round but it doesn't match. Looks like something funny is
+            // going on -- abort.
+            return Err(Error::VerificationFailed(anyhow!(
+                "header does not match previously seen header for the same round"
+            )));
+        }
+
+        // Verify that the state root matches.
+        let roothash_state = RoothashState::new(&state);
+        let state_root = roothash_state
+            .state_root(Context::background(), self.trust_root.runtime_id)
+            .map_err(|err| {
+                Error::VerificationFailed(anyhow!("failed to retrieve trusted state root: {}", err))
+            })?;
+        if runtime_header.state_root != state_root {
+            return Err(Error::VerificationFailed(anyhow!(
+                "state root mismatch (expected: {} got: {})",
+                state_root,
+                runtime_header.state_root
+            )));
+        }
+
+        // Cache verified runtime header.
+        cache
+            .verified_state_roots
+            .put(runtime_header.round, state_root);
+        cache.last_verified_round = runtime_header.round;
+
+        Ok(state)
+    }
+
+    fn trust(&self, cache: &mut Cache, header: ComputeResultsHeader) -> Result<(), Error> {
+        if let Some(state_root) = header.state_root {
+            cache.verified_state_roots.put(header.round, state_root);
+            cache.last_verified_round = header.round;
+        }
+
+        Ok(())
+    }
+
+    /// Start the verifier in a separate thread.
+    pub fn start(self) {
+        std::thread::spawn(move || {
+            let logger = get_logger("consensus/tendermint/verifier");
+            info!(logger, "Starting consensus verifier");
+
+            for retry in 1..=MAX_INITIALIZATION_RETRIES {
```

Comment:
Can you add a comment on why is this retrying needed? Should the thread `JoinHandle` be returned from the `start` method, so that the caller can catch this failing?

Reply:
Retry is needed because `start()` is called when processing the `RuntimeInfoRequest` which is part of the RHP initialization. The host will reject any requests made by the runtime until `RuntimeInfoRequest` is processed successfuly, which leads to this sometimes starting too quickly. I will add a comment.

The problem with a join handle is that it is not obvious where to block for the thread to finish as the protocol thread must not do that. Note that if this fails, then all further verifications will fail (due to the channel receiver being dropped) which will cause the dispatcher to panic and the runtime to abort.

Reply:
Will also limit the restarts to specific error variants as once we're past creating an instance any failures should immediately be fatal.

Reply:
Done.

Reply:
Thanks!